feat: support storing, retrieving and refreshing encrypted secrets at account or asset level (#2236)
FlexMeasures/flexmeasures: documentation/plugin/storing-platform-secrets.rst

        SecretsEncryptor,
        get_secret,
        redact_secrets,
        set_secret,
    )

    encryptor = SecretsEncryptor.from_current_app()

    my_account.secrets = set_secret(
        my_account.secrets,
        "3rdparty-platform.refresh_token",
        refresh_token,
        encryptor=encryptor,
        metadata={"expires_at": refresh_token_expires_at.isoformat()},
    )

    refresh_token = get_secret(
        my_account.secrets,
        "3rdparty-platform.refresh_token",
        encryptor=encryptor,
    )

    response_payload = redact_secrets(my_account.secrets)

The encrypted values are protected by
``FLEXMEASURES_SECRETS_ENCRYPTION_KEY``. This setting accepts an arbitrary
non-empty string, which FlexMeasures derives into a Fernet-compatible key. In
development and tests, FlexMeasures can fall back to ``SECRET_KEY``. Production
installations must set a dedicated encryption key so session signing can be
rotated independently from stored connection credentials.

More details and best practices for storing connection secrets are in the :ref:`connection_secrets_dev` section.

Initiate token refresh cycles on startup
-----------------------------------------

> Most providers will require true credentials only in the first interaction:

Reviewer: From this point on you're discussing fresh tokens specifically. Felt like an unannounced topic switch/zoom.

Author: I added this section at a later point. Will move to the end. And also rewrite for more clarity.

username & password to get the access & refresh token, for example.

> From then, on the refresh token helps to get by, as long as it does not expire.

Reviewer: ```suggestion
From then on, the refresh token helps to get by, as long as it does not expire.
```

A CLI command can be written for this login and seeding of the token refresh cycle.
If no current token is in the database, let your plugin code fail explicitly by

> requiring the user to call this CLI command.

Reviewer: This part I didn't understand.

Token lifecycle strategies
---------------------------

External platforms do not all follow the same token refresh semantics. Keep the
provider-specific HTTP calls in your plugin, and use FlexMeasures utilities only
for encryption, redaction and updating stored token state.

> The following token lifecycle patterns are supported by

Reviewer: I think the terms *token lifecycle strategies*, *token refresh semantics* and *token lifecycle patterns* are used to describe the same thing, increasing cognitive overhead.

Author: Very true, homogenized and explained better now.

        TokenRefreshResult,
        apply_token_refresh_result,
        get_secret,
    )

    encryptor = SecretsEncryptor.from_current_app()

    refresh_token = get_secret(
        my_account.secrets,
        "3rdparty-platform.refresh_token",
        encryptor=encryptor,
    )

    token_response = refresh_with_external_platform(refresh_token)

    my_account.secrets = apply_token_refresh_result(
        my_account.secrets,
        "3rdparty-platform",
        TokenRefreshResult(
            access_token=token_response.get("access_token"),
            refresh_token=token_response.get("refresh_token"),
            access_token_expires_at=token_response.get("access_token_expires_at"),
            refresh_token_expires_at=token_response.get("refresh_token_expires_at"),
            token_type=token_response.get("token_type"),
            metadata={"scope": token_response.get("scope")},
        ),
        encryptor=encryptor,
    )

If a platform only extends the lifetime of the existing access token, keep
``access_token`` set to ``None`` and provide the new
``access_token_expires_at``. The existing encrypted token is kept, and only its
metadata is updated.

> For multi-worker deployments, plugins should cache short-lived access tokens in

Reviewer: "cache"?

``secrets`` with an ``expires_at`` value and refresh them before they expire. A
provider-specific helper can use a database lock around refresh work so only one

> worker refreshes a token while other workers reuse the updated token state.

Reviewer: This reads like a follow-up idea. Not really worked out, yet.

Since the token is stored in the db, the workers would actually be dealing with db transactions, right?

Author: I agree that the db transactions should deal with this okay. I'll remove it.

Author: Well, it could be that worker A needs to wait for the platform response, and in the meantime worker B also requests a freh token (as worker A did not create any sort of db lock yet).

Then, worker B would overwrite the token worker A used. 
In very high-traffic moments, this could happen a lot.

Now, it is up to the use case if these high-traffic moments will ever be expected. If we have a gateway integration and manage hundreds of assets through it, each with its own worker & job, we might get there. But it is something we could also handle in a follow-up.

The docs saying "use a database lock around refresh work" are not super helpful at this time, though.

.. _storing_connection_secrets:

Storing connection secrets
==========================

> Plugins which connect FlexMeasures accounts or assets to external platforms often

Reviewer: ```suggestion
Plugins that connect FlexMeasures accounts or assets to external platforms often
```

        set_secret,
    )

    encryptor = SecretsEncryptor.from_current_app()

    my_account.secrets = set_secret(
        my_account.secrets,
        "3rdparty-platform.refresh_token",
        refresh_token,
        encryptor=encryptor,
        metadata={"expires_at": refresh_token_expires_at.isoformat()},
    )

    refresh_token = get_secret(
        my_account.secrets,
        "3rdparty-platform.refresh_token",
        encryptor=encryptor,
    )

    response_payload = redact_secrets(my_account.secrets)

The encrypted values are protected by
``FLEXMEASURES_SECRETS_ENCRYPTION_KEYS``. This setting accepts arbitrary
non-empty strings, which FlexMeasures derives into Fernet-compatible keys.
Hosts must configure this keyring before secrets can be stored
- FlexMeasures will print a warning if it is not set and hints how to initialize it.

More details and best practices for storing connection secrets are in the :ref:`connection_secrets_dev` section.

Token lifecycle strategies
-----------------------------

External platforms do not all require the same interactions to maintain a connection.
Initial login, token refresh and token expiry can all work rather differently.

> You could say, the implement different "token lifecycle strategies".

Reviewer: ```suggestion
You could say that they implement different "token lifecycle strategies".
```

> That's why our advice is t keep the provider-specific HTTP calls in your plugin,

Reviewer: ```suggestion
That's why our advice is to keep the provider-specific HTTP calls in your plugin,
```

The existing encrypted token is kept, and only its metadata is updated.

Your plugin should also decide when to request a refreshed access token, e.g.:

.. code-block:: python

    3RDPARTY_PLATFORM_TOKEN_LEEWAY = timedelta(seconds=120)
    # you might get this info with `secret_utils.get_secret()`
    if current_access_token_expires_at > now + 3RDPARTY_PLATFORM_TOKEN_LEEWAY:
        return access_token
    response = send_request_to_3rdparty_platform()
    access_token = response.text
    refresh_account.secrets = apply_token_refresh_result(...)

.. _initiating_connection_tokens:

Initiating tokens (before app startup)
-----------------------------------------

Most providers will require the true credentials only in the first interaction:
For example: username & password to get the access & refresh token.
From then on, the refresh token helps to get by (as long as it does not expire).

In your plugin, you can write a CLI command to perform this login, get your first
refresh token and save it as a secret (see `set_secret()` and `apply_token_refresh_result()`
in utils/secrets_utils.py).
The token lifecycle strategy (see above) will depend on the platform you connect to.

> Also advisable: If no current token is in the database, let your plugin code fail

Reviewer: ```suggestion
Also advisable: if no current token is in the database, let your plugin code fail
```

explicitly and advise the user to call your login CLI command.

Alternatively, you can manually store a known credential: Use ``flexmeasures edit secret`` with
an account or asset ID, a dot-separated secret path and either ``--value`` or

> ``--prompt`` (to paste the secret insted of typing it).

Reviewer: ```suggestion
``--prompt`` (to paste the secret instead of typing it).
```

.. _storing_connection_secrets:

Storing connection secrets
==========================

Plugins that connect FlexMeasures accounts or assets to external platforms often
need to store credentials, refresh tokens, access tokens or connection-specific
passwords. Store such values in the ``secrets`` JSON field of the relevant
account or asset, rather than in ``attributes`` or plugin configuration files.

Secrets live in a dictionary and support one or two path levels. If you store
more than one secret per connected platform, identify them by distinct names, for example:

.. code-block:: json

    {
        "3rdparty-platform": {
            "refresh_token": "encrypted-value",

>             "access_token": "encrypted-value",

Reviewer: Are these both a "to-be-encrypted-value"? That is, does this example show something the users would pass or does it show how it would be stored in the db?

Author: Clarified.

>             "access_token_expires_at": "2024-06-01T12:00:00Z"

Reviewer: Wouldn't this be metadata, separately for each field? Or is `"_expires_at"` a special suffix?

Author: Correct. Removed here, the next section shows a nice example.

        }
    }

The ``secrets`` field is intended to be write-only from API and UI flows: users
can provide or replace secret values, but normal responses should return only
redacted metadata such as whether a value is set and when it expires. Trusted
server-side plugin code can decrypt and use the value when it performs work for
the account or asset.

Use ``flexmeasures.utils.secrets_utils`` for secret handling:

.. code-block:: python

    from flexmeasures.utils.secrets_utils import (
        SecretsEncryptor,
        get_secret,
        redact_secrets,
        set_secret,
    )

    encryptor = SecretsEncryptor.from_current_app()

    my_account.secrets = set_secret(
        my_account.secrets,
        "3rdparty-platform.refresh_token",
        refresh_token,
        encryptor=encryptor,

>         metadata={"expires_at": refresh_token_expires_at.isoformat()},

Reviewer: Above, you appear to use a `"3rdparty-platform.access_token_expires_at"` field. But here, you use metadata to indicate an expiration date, for the refresh token instead of the access token. It's not clear to me how these examples relate.

.. _storing_connection_secrets:

Storing connection secrets
==========================

Plugins that connect FlexMeasures accounts or assets to external platforms often
need to store credentials, refresh tokens, access tokens or connection-specific
passwords. Store such values in the ``secrets`` JSON field of the relevant
account or asset, rather than in ``attributes`` or plugin configuration files.

Secrets live in a dictionary and if you store more than one secret per connected platform,
they are indentified by a dot-separated path, for example:

.. code-block:: json

    {
        "3rdparty-platform": {
            "refresh_token": "encrypted-value",
            "access_token": "encrypted-value",
            "access_token_expires_at": "2024-06-01T12:00:00Z"
        }

>     }

Reviewer: There is no dot in this example.

Author: already fixed.
